File: mcp/tools/serena/test/resources/repos/erlang/test_repo/src/services.erl
%% Services module implementing gen_server behavior
-module(services).
-behaviour(gen_server).
-include("../include/records.hrl").
-include("../include/types.hrl").

%% API exports
-export([
    start_link/0,
    stop/0,
    register_user/4,
    get_user/1,
    create_order/2,
    update_order_status/2,
    get_statistics/0
]).

%% gen_server callbacks
-export([
    init/1,
    handle_call/3,
    handle_cast/2,
    handle_info/2,
    terminate/2,
    code_change/3
]).

%% State record
-record(state, {
    users = #{} :: map(),
    orders = #{} :: map(),
    next_user_id = 1 :: integer(),
    next_order_id = 1 :: integer()
}).

%%%===================================================================
%%% API
%%%===================================================================

-spec start_link() -> {ok, pid()} | ignore | {error, term()}.
start_link() ->
    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).

-spec stop() -> ok.
stop() ->
    gen_server:stop(?MODULE).

-spec register_user(string(), email(), integer(), boolean()) -> {ok, user_id()} | {error, term()}.
register_user(Name, Email, Age, Active) ->
    gen_server:call(?MODULE, {register_user, Name, Email, Age, Active}).

-spec get_user(user_id()) -> {ok, #user{}} | {error, not_found}.
get_user(UserId) ->
    gen_server:call(?MODULE, {get_user, UserId}).

-spec create_order(user_id(), list()) -> {ok, integer()} | {error, term()}.
create_order(UserId, Items) ->
    gen_server:call(?MODULE, {create_order, UserId, Items}).

-spec update_order_status(integer(), atom()) -> ok | {error, term()}.
update_order_status(OrderId, Status) ->
    gen_server:call(?MODULE, {update_order_status, OrderId, Status}).

-spec get_statistics() -> #{atom() => integer()}.
get_statistics() ->
    gen_server:call(?MODULE, get_statistics).

%%%===================================================================
%%% gen_server callbacks
%%%===================================================================

-spec init([]) -> {ok, #state{}}.
init([]) ->
    {ok, #state{}}.

-spec handle_call(term(), {pid(), term()}, #state{}) -> handle_call_result().
handle_call({register_user, Name, Email, Age, Active}, _From, State) ->
    UserId = State#state.next_user_id,
    User = models:create_user(UserId, Name, Email, Age),
    UpdatedUser = models:update_user(User, [{active, Active}]),
    NewUsers = maps:put(UserId, UpdatedUser, State#state.users),
    NewState = State#state{
        users = NewUsers,
        next_user_id = UserId + 1
    },
    {reply, {ok, UserId}, NewState};

handle_call({get_user, UserId}, _From, State) ->
    case maps:get(UserId, State#state.users, not_found) of
        not_found -> {reply, {error, not_found}, State};
        User -> {reply, {ok, User}, State}
    end;

handle_call({create_order, UserId, Items}, _From, State) ->
    case maps:get(UserId, State#state.users, not_found) of
        not_found ->
            {reply, {error, user_not_found}, State};
        _User ->
            OrderId = State#state.next_order_id,
            Order = models:create_order(OrderId, UserId, Items),
            NewOrders = maps:put(OrderId, Order, State#state.orders),
            NewState = State#state{
                orders = NewOrders,
                next_order_id = OrderId + 1
            },
            {reply, {ok, OrderId}, NewState}
    end;

handle_call({update_order_status, OrderId, Status}, _From, State) ->
    case maps:get(OrderId, State#state.orders, not_found) of
        not_found ->
            {reply, {error, order_not_found}, State};
        Order ->
            UpdatedOrder = Order#order{status = Status},
            NewOrders = maps:put(OrderId, UpdatedOrder, State#state.orders),
            NewState = State#state{orders = NewOrders},
            {reply, ok, NewState}
    end;

handle_call(get_statistics, _From, State) ->
    Stats = #{
        total_users => maps:size(State#state.users),
        total_orders => maps:size(State#state.orders),
        next_user_id => State#state.next_user_id,
        next_order_id => State#state.next_order_id
    },
    {reply, Stats, State};

handle_call(_Request, _From, State) ->
    {reply, {error, unknown_request}, State}.

-spec handle_cast(term(), #state{}) -> {noreply, #state{}}.
handle_cast(_Msg, State) ->
    {noreply, State}.

-spec handle_info(term(), #state{}) -> {noreply, #state{}}.
handle_info(_Info, State) ->
    {noreply, State}.

-spec terminate(term(), #state{}) -> ok.
terminate(_Reason, _State) ->
    ok.

-spec code_change(term() | {down, term()}, #state{}, term()) -> {ok, #state{}}.
code_change(_OldVsn, State, _Extra) ->
    {ok, State}.

File: mcp/tools/serena/test/resources/repos/elixir/test_repo/lib/services.ex
defmodule TestRepo.Services do
  @moduledoc """
  Services module demonstrating function usage and dependencies.
  Similar to Python's services.py, this module uses the models defined in TestRepo.Models.
  """

  alias TestRepo.Models.{User, Item, Order, Serializable}

  defmodule UserService do
    use GenServer

    # Client API

    @doc """
    Starts the UserService GenServer.
    """
    def start_link(opts \\ []) do
      GenServer.start_link(__MODULE__, %{}, opts)
    end

    @doc """
    Creates a new user and stores it.
    """
    def create_user(pid, id, name, email, roles \\ []) do
      GenServer.call(pid, {:create_user, id, name, email, roles})
    end

    @doc """
    Gets a user by ID.
    """
    def get_user(pid, id) do
      GenServer.call(pid, {:get_user, id})
    end

    @doc """
    Lists all users.
    """
    def list_users(pid) do
      GenServer.call(pid, :list_users)
    end

    @doc """
    Deletes a user by ID.
    """
    def delete_user(pid, id) do
      GenServer.call(pid, {:delete_user, id})
    end

    # Server callbacks

    @impl true
    def init(_) do
      {:ok, %{}}
    end

    @impl true
    def handle_call({:create_user, id, name, email, roles}, _from, users) do
      if Map.has_key?(users, id) do
        {:reply, {:error, "User with ID #{id} already exists"}, users}
      else
        user = User.new(id, name, email, roles)
        new_users = Map.put(users, id, user)
        {:reply, {:ok, user}, new_users}
      end
    end

    @impl true
    def handle_call({:get_user, id}, _from, users) do
      case Map.get(users, id) do
        nil -> {:reply, {:error, :not_found}, users}
        user -> {:reply, {:ok, user}, users}
      end
    end

    @impl true
    def handle_call(:list_users, _from, users) do
      user_list = Map.values(users)
      {:reply, user_list, users}
    end

    @impl true
    def handle_call({:delete_user, id}, _from, users) do
      if Map.has_key?(users, id) do
        new_users = Map.delete(users, id)
        {:reply, :ok, new_users}
      else
        {:reply, {:error, :not_found}, users}
      end
    end
  end

  defmodule ItemService do
    use Agent

    @doc """
    Starts the ItemService Agent.
    """
    def start_link(opts \\ []) do
      Agent.start_link(fn -> %{} end, opts)
    end

    @doc """
    Creates a new item and stores it.
    """
    def create_item(pid, id, name, price, category) do
      Agent.get_and_update(pid, fn items ->
        if Map.has_key?(items, id) do
          {{:error, "Item with ID #{id} already exists"}, items}
        else
          item = Item.new(id, name, price, category)
          new_items = Map.put(items, id, item)
          {{:ok, item}, new_items}
        end
      end)
    end

    @doc """
    Gets an item by ID.
    """
    def get_item(pid, id) do
      Agent.get(pid, fn items ->
        case Map.get(items, id) do
          nil -> {:error, :not_found}
          item -> {:ok, item}
        end
      end)
    end

    @doc """
    Lists all items, optionally filtered by category.
    """
    def list_items(pid, category \\ nil) do
      Agent.get(pid, fn items ->
        item_list = Map.values(items)

        case category do
          nil -> item_list
          cat -> Enum.filter(item_list, &Item.in_category?(&1, cat))
        end
      end)
    end

    @doc """
    Deletes an item by ID.
    """
    def delete_item(pid, id) do
      Agent.get_and_update(pid, fn items ->
        if Map.has_key?(items, id) do
          new_items = Map.delete(items, id)
          {:ok, new_items}
        else
          {{:error, :not_found}, items}
        end
      end)
    end
  end

  defmodule OrderService do
    @doc """
    Creates a new order.
    """
    def create_order(id, user, items \\ []) do
      Order.new(id, user, items)
    end

    @doc """
    Adds an item to an existing order.
    """
    def add_item_to_order(order, item) do
      Order.add_item(order, item)
    end

    @doc """
    Updates the status of an order.
    """
    def update_order_status(order, status) do
      Order.update_status(order, status)
    end

    @doc """
    Processes an order (changes status to :processing).
    """
    def process_order(order) do
      update_order_status(order, :processing)
    end

    @doc """
    Completes an order (changes status to :completed).
    """
    def complete_order(order) do
      update_order_status(order, :completed)
    end

    @doc """
    Cancels an order (changes status to :cancelled).
    """
    def cancel_order(order) do
      update_order_status(order, :cancelled)
    end
  end

  @doc """
  Factory function to create a service container.
  """
  def create_service_container do
    {:ok, user_service} = UserService.start_link()
    {:ok, item_service} = ItemService.start_link()

    %{
      user_service: user_service,
      item_service: item_service,
      order_service: OrderService
    }
  end

  @doc """
  Helper function to serialize any model that implements the Serializable protocol.
  """
  def serialize_model(model) do
    Serializable.to_map(model)
  end

  # Module-level variables for testing
  @sample_user_id "sample_user"
  @sample_item_id "sample_item"

  @doc """
  Gets the sample user ID.
  """
  def sample_user_id, do: @sample_user_id

  @doc """
  Gets the sample item ID.
  """
  def sample_item_id, do: @sample_item_id

  # Create some sample data at module load time
  def setup_sample_data(container) do
    # Create sample user
    UserService.create_user(
      container.user_service,
      @sample_user_id,
      "Sample User",
      "sample@example.com",
      ["user", "customer"]
    )

    # Create sample item
    ItemService.create_item(
      container.item_service,
      @sample_item_id,
      "Sample Widget",
      29.99,
      "electronics"
    )
  end
end 